crates/cairo-profiler/src/trace_reader/function_trace_builder.rs

```diff
@@ -220,3 +251,62 @@ fn build_current_call_stack(
         current_call_stack
     }
 }
+
+fn stack_trace_to_samples(
+    functions_stack_traces: HashMap<Vec<FunctionCall>, Steps>,
+    os_resources_map: &OsResources,
+) -> Vec<Sample> {
+    functions_stack_traces
+        .into_iter()
+        .map(|(call_stack, steps)| {
+            let mut measurements: HashMap<MeasurementUnit, MeasurementValue> = vec![(
+                MeasurementUnit::from("steps".to_string()),
+                MeasurementValue(i64::try_from(steps.0).unwrap()),
+            )]
+            .into_iter()
+            .collect();
+
+            if let Some(FunctionCall::InternalFunctionCall(InternalFunctionCall::Syscall(
+                function_name,
+            ))) = call_stack.last()
+            {
+                let Ok(syscall) = map_syscall_name_to_selector(function_name.0.as_str()) else {
+                    // todo: print the error in debug mode
+                    return Sample {
+                        call_stack,
+                        measurements,
+                    };
+                };
+                let resources = os_resources_map
+                    .execute_syscalls
+                    .get(&syscall)
+                    .unwrap_or_else(|| {
+                        panic!("Missing syscall {syscall:?} from versioned constants file")
+                    });
+
+                if let Some(value) =
+                    measurements.get_mut(&MeasurementUnit::from("steps".to_string()))
+                {
+                    *value += MeasurementValue(i64::try_from(resources.n_steps).unwrap());
+                }
+
+                measurements.insert(
+                    MeasurementUnit::from("memory_holes".to_string()),
+                    MeasurementValue(i64::try_from(resources.n_memory_holes).unwrap()),
+                );
+
+                for (builtin, b_count) in &resources.builtin_instance_counter {
+                    measurements.insert(
+                        MeasurementUnit::from(builtin.to_string()),
+                        MeasurementValue(i64::try_from(*b_count).unwrap()),
+                    );
+                }
```

Comment:
A heads up so we don't break it in the future when we have builtins and memory holes counting during function level trace collecting. This code snippet requires `impl From<&str> for MeasurementUnit`, `impl TryFrom<usize> for MeasurementValue` and `#[derive(Default)]` on `MeasurementValue`.

```suggestion
                *measurements.entry("memory_holes".into()).or_default() +=
                    resources.n_memory_holes.try_into().unwrap();

                for (builtin, b_count) in &resources.builtin_instance_counter {
                    *measurements.entry(builtin.as_ref().into()).or_default() +=
                       (*b_count).try_into().unwrap();
                }
```

Reply:
Is it okay if I implement this in #78 ? There is a chance I will modify it even further during this task

Reply:
Yep, it's my pedantism, feel free to skip it for now